Watch improvements

eatkins opened the pull request:
- [x] I've read the [CONTRIBUTING](https://github.com/sbt/sbt/blob/1.x/CONTRIBUTING.md) guidelines


This is a large pull request that is the culmination of a number of month's work trying to improve the overall io and watch performance of sbt. For the most part, I've broken the commits up so that each one more or less stands on its own. I am doing it as one large pull request just because the overhead of making 20 partially dependent commits feels to high. It is linked with https://github.com/sbt/io/pull/177.

At a high level, this PR does two things:

1) Improve continuous builds. I made it possible to add custom trigger rules so that the user can, for example, cancel the watch if the command fails. I also fixed logging so that it should be easier to debug any bugs in the field.

2) Add a global in memory cache of the file system to improve incremental compilation start up times. This works well to reduce the startup time of incremental compilation. For a simple project with 5000 files, I've seen the compilation startup time drop from O(1 second) to O(200ms) (for small projects the startup time is O(50ms) where the 50ms is entirely spent in the update task to re-resolve the project). To achieve this speed up we combine the cache with the ExternalLookup in zinc so that we can bypass the io cost of hashing the source file contents. I also add a hybrid option that polls some directories and uses the regular os notifications for the remainder. This is to handle the case where, say, inotify isn't supported on a particular file system mount point (I think this may be the underlying issue with #4179 ). 

I repeated the experiment in [mill watch latency](http://www.lihaoyi.com/post/MillBetterScalaBuilds.html#watch--re-evaluate). With a warm jvm and if I use a resident compiler, I see O(120ms) latency between when file is modified and when _compilation_ is **complete** not when it begins. Also, in my test project with 5000 files mill burns an entire cpu polling the file system while sbt sits at about 0.3% of a cpu.

Seeing as this is a large change, I expect that there may be a good amount of discussion. Wherever possible, I've tried to keep the user visible impact of this change to a minimum. Moreover, everything I've added is configurable in such a way that a user should be able to revert back to the existing behavior with a few simple modifications if there are any new issues. It is always scary to add a global cache. In this case, I think the risk is not that bad both because I added a task to flush its contents and touching a monitored directory also triggers a full rescan. At any rate, users can choose to eschew the cache via the fileTreeViewConfig key.

I also am still working on adding some more scripted tests of continuous builds. I wanted to get this out the door, but I do expect to update a few commits. I don't expect to change much of the mainline code though unless errors are discovered writing additional scripted tests.

eatkins committed:
Deprecate unused method

eatkins committed:
Bump watch anti-entropy

There have been reports that often a new build will be triggered
immediately after the previous build even when none of the files have
been modified since the start of the last build. This can happen when,
for example, a program implements save with a rename. When that occurs,
a deletion watch event may trigger the build but the corresponding
creation event may be detected outside of the current 40ms window. By
bumping this value to 500ms, we hopefully prevent the majority of these
false triggers. For unusual workflows in which this longer quarantine
period is an issue, the setting can be overridden.

eatkins committed:
Refactor continuous execution

This commit makes watch event logging work in the '~' command. The
previous design of the command made this difficult, so there is a
significant re-design of the implementation of '~'. I believe that this
redesign will allow the feature to be maintained and improved more
easily moving forward. With the redesign, it is now possible to test the
business logic of the watch command (and I add a rudimentary test that I
will build upon in subsequent commits).

A bonus of this redesign is that now if the user tries to watch an
invalid command, the watch will immediately terminate with an error
rather than get stuck waiting for events when the task can never
possibly succeed.

The previous implementation of the '~' command makes it difficult
to dynamically control the implementation arguments because it is
implemented in the command project which makes it unable to depend on
any task keys that are defined in the build. It works around this by
putting all of it's configuration in the Watched attribute which is
stored globally. This would not have been necessary if the function had
been defined in the main project where it could just extract the value
of the watched task rather than relying on the global attribute value.
Moreover, because it cannot depend on tasks, it makes it nigh impossible
to use the logging framework within the '~' command.

Another issue with the previous implementation is that it's somewhat
difficult to reason about. The executeContinuously has effectively two
entry points: one for the first time the command is run and one for each
subsequent invocation when a new build is triggered. The successive
invocations are triggered by prepending commands to run to the previous
state. This is made recursive by prepending the initial command (that
was prefixed with '~'. Which branch we're in is determined by checking
for the existence of a temporary attribute, that we must ensure that we
remove when the build is stopped. This makes a lot of behavior non-local and
difficult for an outsider who is less familiar with sbt to understand.

Broadly, this refactor does two things:
1) Move the definition of continuous from BasicCommands to BuiltInCommands
2) Re-work the implementation to be executed in code rather than using
   the sbt dsl.

The first part is simple. We just add an implementation of continuous to
BuiltInCommands and remove it from the list of BasicCommands. We need to
leave in the legacy implementation for binary compatibility. I also
moved all of the actual implementation logic into Watched, which makes
maintenance easier since most of the logic is in one place.

The second part is more complicated. Rather than rely on the sbt dsl
(e.g. `(ClearOnFailure :: next :: FailureWall :: repeat :: s)`) to
parse and run the command. We manually parse the command and generate a
task of type `() => State`. We don't actually need to do anything with
the generated state because we're going to return the original state at
the end of the command no matter what. With this task, we can then
create a tail recursive function that repeatedly executes the task until
the watch is terminated.

The parsing is handled in the Watch.command method (which is where I
moved the refactored BasicCommands.continuous implementation). The
actual task running and monitoring is handled in Watched.watch. This
method has no reference to the sbt state, which makes it testable. It sets
up an event monitor and then delegates the recursive monitoring to a
small nested function, Watched.watch.impl. One nice thing about this
approach is that it is very easy to reason about the life cycle of the
EventMonitor. The recursive call is within a try { } finally { } where
the monitor and stdin are guaranteed to be cleared at the end.

Adding support for a custom (and default) watch logger is trivial with
the new infrastructure and is done via the watchLogger TaskKey.

There was a small reporting race condition that was introduced by the
change to (2). Because the new implementation is able to bypass command
parsing for triggered builds, the watch message would usually end up
being printed before the task outcome was fully logged. To work around
this, I made the watch and triggered messages be logged rather than
printed directly to stdout. As a result, the only user visible result of
this change should be that instead of seeing:
"1. Waiting for source changes in project foo... (press enter to interrupt)",
users will now see:
"[info] 1. Waiting for source changes in project foo... (press enter to interrupt)".

eatkins committed:
Do not append default watch sources with ++=

Using ++= prevents these from being overridden by the user without doing
some nasty runtime manipulations.

eatkins committed:
Remove intellij warnings

eatkins committed:
Move executeContinuously into LegacyWatched

This helps keep Watched.scala more manageable until we can potentially
remove this code.

eatkins committed:
Set swoval.tmpdir property on startup

This change makes the temporary shared library that is created by
the swoval file-tree-views library to be extracted into the sbt global
base directory rather than the temp file. This way if there is a leak of
shared libraries, they can easily be found in ~/.sbt rather than in,
say, /tmp (or the osx/windows equivalent location). The extracted shared
library objects will be in the path ~/.sbt/swoval-jni. There is a
shutdown hook that removes them as well as a garbage collection process
that runs in the background whenever the swoval library is loaded, so
these shouldn't leak uncontrollably.

eatkins committed:
Add custom external hooks

For projects with a large number of files, zinc has to do a lot of work
to determine which source files and binaries have changes since the last
build. In a very simple project with 5000 source files, it takes roughly
750ms to do a no-op compile using the default incremental compiler
options. After this change, it takes about 200ms. Of those 200ms, 50ms
are due to the update task, which does a partial project resolution*.

The implementation is straightforward since zinc already provides an api
for overriding the built in change detection strategy. In a previous
commit, I updated the sources task to return StampedFile rather than
regular java.io.File instances. To compute all of the source file
stamps, we simply list the sources and if the source is in fact an
instance of StampedFile, we don't need to compute it, otherwise we
generate a StampedFile on the fly. After building a map of stamped files
for both the sources files and all of the binary dependencies, we simply
diff these maps with the previous results in the changedSources,
changedBinaries and removedProducts methods.

The new ExternalHooks are easily disabled by setting
`externalHooks := _ => None`
in the project build.

In the future, I could see moving ExternalHooks into the zinc project so
that other tools like bloop or mill could use them.

* I think this delay could be eliminated by caching the UpdateResult so
long as the project doesn't depend on any snapshot libraries. For a
project with a single source, the no-op compile takes O(50ms) so caching
the project resolution would make compilation start nearly
instantaneous.

eatkins committed:
Add global file repository task

Every time that the compile task is run, there are potentially a large
number of iops that must occur in order for sbt to generate the source
file list as well as for zinc to check which files have changed since
the last build. This can lead to a noticeable delay between when a build
is started (either manually or by triggered execution) and when
compilation actually begins. To reduce this latency, I am adding a
global view of the file system that will be stored in
BasicKeys.globalFileTreeView.

To make this work, I introduce the StampedFile trait, which augments the
java.io.File class with a stamp method that returns the zinc stamp for
the file. For source files, this will be a hash of the file, while for
binaries, it is just the last modified time. In order to gain access to
the sbt.internal.inc.Stamper class, I had to append addSbtZinc to the
commandProj configurations.

This view may or may not use an in-memory cache of the file system tree
to return the results. Because there is always the risk of the cache
getting out of sync with the actual file system, I both make it optional
to use a cache and provide a mechanism for flushing the cache. Moreover,
the in-memory cache implementation in sbt.io, which is backed by a
swoval FileTreeRepository, has the property that touching a monitored
directory invalidates the entire directory within the cache, so the
flush command isn't even strictly needed in general.

Because caching is optional, the global is of a FileTreeDataView, which
doesn't specify a caching strategy. Subsequent commits will make use of
this to potentially speed up incremental compilation by caching the
Stamps of the source files so that zinc does not need to compute the
hashes itself and will allow for continuous builds to use the cache to
monitor events instead of creating a new, standalone FileEventMonitor.

eatkins committed:
Use FileTreeDataView to collect files

Now that we have the fileTreeView task, we can generalized the process
of collecting files from the view (which may or may not actually cache
the underlying file tree). I moved the implementation of collectFiles
and addBaseSources into the new FileManagement object because Defaults
is already too large of a file. When we query the view, we also need to
register the directory we're listing because if the underlying view is a
cache, we must call register before any entries will be available.
Because FileTreeDataView doesn't have a register method, I implement
registration with a simple implicit class that pattern matches on the
underlying type and only calls register if it is actually a
FileRepository.

A side effect of this change is that the underlying files returned by
collectFiles and appendBaseSources are StampedFile instances. This is so
that in a subsequent commit, I can add a Zinc external hook that will
read these stamps from the files in the source input array rather than
compute the stamp on the fly. This leads to a substantial reduction in
Zinc startup time for projects with many source files. The file filters
also may be applied more quickly because the isDirectory property (which
we check for all source files) is read from a cached value rather than
requiring a stat.

I had to update a few of the scripted tests to use the `1.2.0`
FileTreeViewConfig because those tests would copy a file and then
immediately re-compile. The latency of cache invalidation is O(1-10ms),
but not instantaneous so it's necessary to either use a non-caching
FileTreeView or add a sleep between updates and compilation. I chose the
former.

eatkins committed:
Add support for polling some directories

It is not always possible to monitor a directory using OS file system
events. For example, inotify does not work with nfs. To work around
this, I add support for a hybrid FileTreeViewConfig that caches a
portion of the file system and monitors it with os file system
notification, but that polls a subset of the directories. When we query
the view using list or listEntries, we will actually query the file
system for the polling directories while we will read from the cache for
the remainder. When we are not in a continuous build (~ *), there is no
polling of the pollingDirectories but the cache will continue to update
the regular directories in the background. When we are in a continuous
build, we use a PollingWatchService to poll the pollingDirectories and
continue to use the regular repository callbacks for the other
directories.

I suspect that #4179 may be resolved by adding the directories for which
monitoring is not working to the pollingDirectories task.

eed3si9n commented on build.sbt:
```diff
@@ -9,6 +9,7 @@ import local.Scripted
 def buildLevelSettings: Seq[Setting[_]] =
   inThisBuild(
     Seq(
+      doc in Compile := file("/dev/null"),
```
What's the purpose of this?

eatkins replied:
https://github.com/sbt/sbt/pull/4335/commits/ba6d0fb4a7d799f6c1011be199a36370414090ac should be reverted. I was using this to speed up local development but didn't mean to include it in the PR.

eed3si9n commented on main-command/src/main/scala/sbt/FileTreeViewConfig.scala:
```diff
@@ -0,0 +1,164 @@
+   * @param antiEntropy the duration of the period after a path triggers a build for which it is
+   * @return a [[FileTreeViewConfig]] instance.
+   */
+  def `1.2.0`(pollingInterval: FiniteDuration, antiEntropy: FiniteDuration): FileTreeViewConfig =
```
Not sure if this name is clear when it comes up as 

```scala
fileTreeViewConfig := FileTreeViewConfig.`1.2.0`(pollInterval.value, watchAntiEntropy.value)
```

Maybe `sbt1_2_compat`?

eed3si9n commented on main-command/src/main/scala/sbt/Watched.scala:
```diff
@@ -103,8 +112,127 @@ object Watched {
-  def printIfDefined(msg: String) = if (!msg.isEmpty) System.out.println(msg)
+  def printIfDefined(msg: String): Unit = if (!msg.isEmpty) System.out.println(msg)
+
+  type Task = () => State
```
The term task mean something already so please pick some other term here.
It would be good to add Scaladoc comment on what it is.

eatkins replied:
Fair point. I might also be able to remove this alias anyway. I think I was mainly using it so that I could avoid the extra parens in `type Setup = (State, Watched, String) => (State, Logger, (() => State))`

eed3si9n commented on main-command/src/main/scala/sbt/Watched.scala:
```diff
@@ -103,8 +112,127 @@ object Watched {
+  def printIfDefined(msg: String): Unit = if (!msg.isEmpty) System.out.println(msg)
+
+  type Task = () => State
+  type Setup = (State, Watched, String) => (State, Logger, Task => State)
```
I tend to use the term "configuration" for these types of things.
Or maybe `WatchSetup`.

eatkins replied:
I'll come up with an alternate name. I didn't put a lot of thought into this to be honest.

eed3si9n commented on main-command/src/main/scala/sbt/Watched.scala:
```diff
@@ -103,8 +112,127 @@ object Watched {
+   * @return the initial state if all of the input commands are valid. Otherwise, returns the
+   *         initial state with the failure transformation.
+   */
+  private[sbt] def command(
```
Can we call this `executeContinuously`?

eatkins replied:
Sure.

eed3si9n commented on main-command/src/main/scala/sbt/Watched.scala:
```diff
@@ -148,6 +276,112 @@ object Watched {
+  /**
+   * Provides a number of configuration options for continuous execution.
+   */
+  trait Config {
```
If this is meant to be used by the build user, I think it's better to move it out of `object Watched`, and name it `WatchConfiguration` or something.

eatkins replied:
I guess for me it's a bit of a style thing. I find it intuitive to scope the kind of Configuration by putting it inside of the companion object rather than using semantic scoping with a name prefix. But I don't feel strongly enough to argue against making this change.

eed3si9n commented on main-command/src/main/scala/sbt/Watched.scala:
```diff
@@ -148,6 +276,112 @@ object Watched {
+     * Returns true if the continuous execution should stop.
+     * @return true if the contiuous execution should stop.
+     */
+    def shouldTerminate(): Boolean
```
"should terminate" sounds like it should eventually terminate (finish calculation).
If it's for stopping, maybe flip the boolean and call it "isAlive"?

eatkins replied:
Good point. I'd reused the previous method name, but you're right that is unclear.

eatkins replied:
I was going back through the PR and realized that this method ends up being removed by the end of the PR anyway. It's just a placeholder to make the git/pr history more readable.

eed3si9n commented on main-command/src/main/scala/sbt/Watched.scala:
```diff
@@ -50,18 +62,46 @@ trait Watched {
 object Watched {
-  val defaultWatchingMessage: WatchState => String = ws =>
-    s"${ws.count}. Waiting for source changes... (press enter to interrupt)"
+  sealed trait Action
```
These probably should be generated using Contraband.

eatkins replied:
Can you expand on that? I understand that Contraband is used to evolve binary compatible apis, but I'm not exactly clear on how that would apply here.

eed3si9n replied:
I use Contraband for enums as well - https://github.com/sbt/sbt/blob/develop/main/src/main/contraband/main.contra
But I just noticed that Action is not completely sealed because of Custom, so that might not be a good fit.

eatkins replied:
Aah, makes sense. I am not opposed to contraband. The `Custom` action is sort of speculative in the event that users might want to use ~ to potentially trigger any arbitrary state transformation. My guess is that very, very few users would ever actually do this (even if I think it's a cool capability).

eed3si9n reviewed (changes requested):
I picked on some superficial things.

eatkins committed:
Add the ability to halt watch on failure

There may be instances where the user may wish to stop the watch if an
error occurs running the task. To facilitate this, I add boolean
parameter, lastStatus, to watchShouldTerminate. The value is computed by
modifying the state used to run the task to have a custom onFailure
command. If the task fails, the returned state will have the onFailure
command will be enqueued at the head of the remaining commands. The
result of the task then becomes true if the custom onFailure is not
present in the remaining commands and false if it is. We don't actually
run this command, so it's just implemented with the identity function.

I also updated Watched.watch to return an Action instead of Unit. This
enables us to return a failed state if Watched.watch returns
HandleError.

eatkins committed:
Add support for user defined Actions

It may be useful for users to be able to return their own custom
Action types in the Config callbacks. For a contrived example, a user
could add a jar file in the .ivy2 directory to the watch sources and
trigger a reboot full when that jar changes.

eed3si9n commented on main-command/src/test/scala/sbt/WatchedSpec.scala:
```diff
@@ -0,0 +1,143 @@
+import scala.concurrent.duration._
+
+class WatchedSpec extends FlatSpec with Matchers {
+<<<<<<< HEAD
```
Git merge artifact

eatkins replied:
Whoops. Fixed.

eatkins committed:
Refactor Watched

This commit reworks Watched to be more testable and extensible. It also
adds some small features. The previous implementation presented a number
of challenges:

1) It relied on external side effects to terminate the watch, which was
   difficult to test
2) It exposed irrelevant implementation details to the user in the
   methods that exposed the WatchState as a parameter.
3) It spun up two worker threads. One was to monitor System.in for user
   input. The other was to poll the watch service for events and write
   them to a queue. The user input thread actually broke '~console'
   because nearly every console session will hit the <enter> key, which
   would eventually cause the watch to stop when the user exited the
   console.

To address (1), I add the shouldTerminate method to WatchConfig. This
takes the current watch iteration is input and if the function returns
true, the watch will stop.

To address (2), I replace the triggeredMessage and watchingMessage keys
with watchTriggeredMessage and watchStartMessage. The latter two keys
are functions that do not take the WatchState as parameters. Both
functions take the current iteration count as a parameter and the
watchTriggeredMessage also has a parameter for the path that triggered
the build.

To address (3), I stop using the sbt.internal.io.EventMonitor and
instead use the sbt.io.FileEventMonitor. The latter class is similar to
the former except that it's polling method accepts a duration, which may
be finite or infinite) and returns all of the events that occurred since
it was last polled. By adding the ability to poll for a finite amount of
time, we can interleave polling for events with polling System.in for
user input, all on the main thread. This eliminates the two extraneous
threads and fixes the '~console' use case I described before.

I also let the user configure the function that reads from System.in via
the watchHandleInput method. In fact, this method need not read from
System.in at all since it's just () => Watched.Action. The reason that
it isn't () => Boolean is that I'd like to leave open the option for the
ability to trigger a build via user input, not just terminating the
watch. My initial idea was to add the ability to type 'r' to re-build in
addition to <enter> to exit. This doesn't work without integrating
jline though because the input is buffered. Regardless, for testing
purposes, it gives us the ability to add a timeout to the watch by
making handleInput return true when a deadline expires.

The tests are a bit wonky because I still need to rely on side effects
in the logging methods to orchestrate the sequence of file events that
I'd like to test. While I could move some of this logic into a
background thread, there still needs to be coordination between the
state of the watch and the background thread. I think it's easier to
reason about when all of the work occurs on the same thread, even if it
makes these user provided functions impure.

I deprecated all of the previous watch related keys that are no longer
used with the new infrastructure. To avoid breaking existing builds, I
make the watchConfig task use the deprecated logging methods if they are
defined in the user's builds, but sbt will not longer set the default
values. For the vast majority of users, it should be straightforward to
migrate their builds to use the new keys.  My hunch is that the of the
deprecated keys, only triggeredMessage is widely used (in conjunction
with the clear screen method) and it is dead simple to replace it with
watchTriggeredMessage.

Note: The FileTreeViewConfig class is not really necessary for this commit.
It will become more important in a subsequent commit which introduces an
optional global file system cache.

eatkins committed:
Add support to rebuild a '~' task by pressing 'r'

Sometimes a user may want to rerun their task even if the source files
haven't changed. Presently this is a little annoying because you have to
hit enter to stop the build and then up arrow or <ctrl+r> plus enter to
rebuild. It's more convenient to just be able to press the 'r' key to
re-run the task.

To implement this, I had to make the watch task set up a jline terminal
so that System.in would be character buffered instead of line buffered.
Furthermore, I took advantage of the NonBlockingInputStream
implementation provided by jline to wrap System.in. This was necessary
because even with the jline terminal, System.in.available doesn't return
> 0 until a newline character is entered. Instead, the
NonBlockingInputStream does provide a peek api with a timeout that will
return the next unread key off of System.in if there is one available.
This can be use to proxy available in the WrappedNonBlockingInputStream.

To ensure maximum user flexibility, I also update the watchHandleInput Key to
take an InputStream and return an Action. This setting will now receive
the wrapped System.in, which will allow the user to create their own
keybindings for watch actions without needing to use jline themselves.

Future work might make it more straightforward to go back to a line
buffered input if that is what the user desires.

eatkins commented:
In addition to making the requested changes from @eed3si9n, I figured out how to use jline to swap stdin to char buffered rather than line buffered which allowed me to add the feature that while in a continuous build, typing 'r' will re-run the command (see 
https://github.com/sbt/sbt/pull/4335/commits/fe64c5e7dc0d2a91ae0827b082a51073c0dddaa5) .

eatkins committed:
Only use a file repository for interactive sessions

I realized that using the cache has the potential to cause issues for
batch processing in CI if some tasks assume that a file created by one
task will immediately be visible in the other. With the cache, there is
typically on O(10ms) latency between a file being created and appearing
in the cache (at least on OSX). When manually running commands, that
latency doesn't matter.

eatkins committed:
Use FileTreeRepository when interactive or continuous

Ideally we use the FileTreeRepository for interactive sessions by
default. A continuous build is effectively interactive, so I'd like that
case to also use the file tree repository. To avoid breaking scripted
tests, many of which implicitly expect file tree changes to be
instantaneously available, we set interactive to true only if we are not
in a scripted run, which can be verified by checking that the commands
contains "setUpScripted".

eatkins commented:
I tacked on 
https://github.com/sbt/sbt/pull/4335/commits/0a9ae7b4b49da95bcac27db8dace663b54ff7686
though I wish that I'd just gone ahead and merged first so that it could have been a separate PR.

dwijnand commented on main/src/main/scala/sbt/Defaults.scala:
```diff
@@ -351,12 +369,14 @@ object Defaults extends BuildCommon {
+        excludeFilter in unmanagedSources
+      )
+      .value,
+    watchSources in ConfigGlobal := (watchSources in ConfigGlobal).value ++ {
```
> Using ++= prevents these from being overridden by the user without doing
some nasty runtime manipulations.

I'm confused by this?  How is this line different than the original?

eatkins replied:
Welcome back, @dwijnand :). I added this line because it was impossible for me to remove the default watchSources in the CloseWatchPlugin because of the use of `++=`. In a plugin or a build, if you do something like `watchSources in ConfigGlobal := Nil`, you will still end up with the default watch sources. I had to manually remove these in the CloseWatchPlugin in a very obscure way: https://github.com/swoval/swoval/blob/0d4882e87e0d1deb710338e3d55bed17d772bbe4/plugin/src/main/scala/com/swoval/watchservice/CloseWatchPlugin.scala#L278

dwijnand replied:
OK, the reason for that is probably because in `Defaults` it's defined in `sourceConfigPaths` (i.e applied per configuration), while in `CloseWatchPlugin` you were probably trying to define it in `globalSettings`.  So it's an ordering thing of which definition in Global scope wins.

But that doesn't 100% explain why `k ++= x` transformed to `k := k.value ++ x` aren't the same.  Unless, of course, they're subtly different when sbt does it's setting re-ordering thing as it processes them (I forget where).

🤷‍♂️

eatkins replied:
You know, I have to admit that I never actually tested this change, so I have no idea if it actually makes a difference. My mental model was just that `:=` always overrides any values set previously with `:=` and that `++=` always appends regardless of any previous overrides of `:=`. But I suspect that is not actually correct given your comments. The guide to scopes in the reference manual is helpful, but I wish there was a more detailed explanation of the exact key resolution algorithm.

dwijnand replied:
I think your intuition is right.  I was just grasping at straws trying to come up with plausible reasons why the change made a difference.  But if you're saying you didn't test it, then I'm more confident that it changed nothing. 😄

`:=` is actually interesting because it'll even suppress setting mistakes, for instance:

```scala
val someSources = settingKey[Seq[String]]("")

someSources ++= Seq("a", "b", "c") // or 'someSources := someSources.value ++ ..'
someSources := Seq("d")
```

results in `Seq("d")`, but without the `:=` line it results in

```scala
[error] Reference to undefined setting:
[error]
[error]   someSources from someSources (/s/t-test-order/build.sbt:4)
```

dwijnand replied:
Oh and large parts of my understand about scopes and settings comes from experimentation 🙂

dwijnand commented on main/src/main/scala/sbt/Defaults.scala:
```diff
@@ -377,6 +397,13 @@ object Defaults extends BuildCommon {
         else Nil
       bases.map(b => new Source(b, include, exclude)) ++ baseSources
     },
+    watchProjectSources in ConfigGlobal := (watchProjectSources in ConfigGlobal).value ++ {
```
Is it sufficient to `Global / watchProjectSources := Nil` to opt out of this?

eatkins replied:
Should be. I am planning on overhauling the documentation before the 1.3.0 release to explain all of this.

dwijnand replied:
I was hoping to hear something like that, as I didn't notice any notes in this PR (which was never a hard requirement).  Lovely, thank you.  You've been doing some amazing stuff.

eatkins replied:
Thanks for the feedback! It's always nice to hear that your work is appreciated.
